### PR TITLE
Implement a tree-sitter scanner for template disambiguation

### PR DESCRIPTION
**Pull request by ben-clayton**
Use a custom scanner to disambiguate between template-argument-lists and less-than / greater-than, as proposed in #3770.

Very rough - but demonstrates a working proof of concept.

### CODE REVIEW

File: wgsl/scanner.cc [hunk 1]
```diff
@@ -0,0 +1,927 @@
...
+inline bool operator<(CodePoint code_point, CodePointRange range) {
+  return code_point < range.first;
+}
+inline bool operator<(CodePointRange range, CodePoint code_point) {
+  return range.last < code_point;
+}
+
+// Interval ranges of all code points in the Unicode 14 XID_Start set
+// This array needs to be in ascending order.
+constexpr CodePointRange kXIDStartRanges[] = {
```
Comment (mehmetoguzderin) [inline wgsl/scanner.cc:93]: I wonder if it is possible to reuse Tree-sitter's internal existing knowledge of this code point range (used for the identifier pattern).

Comment (ben-clayton) [reply wgsl/scanner.cc:93]: Yeah, I wondered that too and mentioned it in my feedback at https://github.com/tree-sitter/tree-sitter/issues/2063.

Comment (mehmetoguzderin) [reply wgsl/scanner.cc:93]: That'd both concise and make it easier to not make scanner part of source tree

Comment (mehmetoguzderin) [reply wgsl/scanner.cc:93]: @ben-clayton would being able to access to `keyword_lex_fn` in TSLanguage provide a solution?

Comment (ben-clayton) [reply wgsl/scanner.cc:93]: No idea. I literally used tree sitter for the first time yesterday.

A possibly easier solution might be to find a dependency we can use that has these unicode tables. I presume tree sitter must be using some table, somewhere.

**Comment by ben-clayton** [comment]
@dneto0 has kindly taken this over.

**Pull request closed without merging**